langchain4j-jlama/src/main/java/dev/langchain4j/model/jlama/JlamaChatModel.java

```diff
@@ -62,22 +65,68 @@ public static JlamaChatModelBuilder builder() {
 
     @Override
     public Response<AiMessage> generate(List<ChatMessage> messages) {
+       return generate(messages, List.of());
+    }
+
+    @Override
+    public Response<AiMessage> generate(List<ChatMessage> messages, List<ToolSpecification> toolSpecifications) {
         if (model.promptSupport().isEmpty())
             throw new UnsupportedOperationException("This model does not support chat generation");
 
-        PromptSupport.Builder promptBuilder = model.promptSupport().get().newBuilder();
+        PromptSupport.Builder promptBuilder = model.promptSupport().get().builder();
+
         for (ChatMessage message : messages) {
             switch (message.type()) {
-                case SYSTEM -> promptBuilder.addSystemMessage(message.text());
-                case USER -> promptBuilder.addUserMessage(message.text());
-                case AI -> promptBuilder.addAssistantMessage(message.text());
+                case SYSTEM -> promptBuilder.addSystemMessage(((SystemMessage)message).text());
+                case USER -> {
+                    UserMessage userMessage = (UserMessage)message;
+                    for (Content content : userMessage.contents()) {
+                        if (content.type() != ContentType.TEXT)
+                            throw new UnsupportedOperationException("Unsupported content type: " + content.type());
+
+                        promptBuilder.addUserMessage(((TextContent)content).text());
```

Comment:
The prompt templates handle this by doing just that but it's upto the model?

Reply:
This code:
```Java
UserMessage userMessage = UserMessage.from(List.of(TextContent.from("hello"), TextContent.from("world")));
model.generate(List.of(userMessage));
```
will create this prompt:
```
<|start_header_id|>system<|end_header_id|>

Environment: ipython
Cutting Knowledge Date: December 2023
Today Date: 26 Jul 2024

<|eot_id|><|start_header_id|>user<|end_header_id|>

Given the following functions, please respond with a JSON for a function call with its proper arguments that best answers the given prompt.

Respond in the format {"name": function name, "parameters": dictionary of argument name and its value}.Do not use variables.

hello<|eot_id|><|start_header_id|>user<|end_header_id|>

world<|eot_id|><|start_header_id|>assistant<|end_header_id|>

```

It seems there are 2 problems:
- 2 user messages instead of 1. I am not sure how exactly other LLM providers handle multiple contents in a single user message, but I guess some form of concatenation hapens. Usually contents are of different type though (e.g. text + image). So this is not really critical, can be reconsidered when/if other content types will be suported.
- Instructions regarding tools are present, but I did not specify any tools, Is this intended?

Reply:
Hmm good catch I pushed a fix for second point.

Regarding the first point the order matters.  Esp for message history being passed in. Meaning I can't concat all user messages together arbitrarily.   
I think you're suggesting I collapse any messages of the same role that are sequential but I think the llm would handle it as is.

Reply:
> Meaning I can't concat all user messages together arbitrarily.

I do not suggest concatenating all user messages together. I suggest concatenating all **_parts_** (contents) of a **_single_** user message together. In some LLM providers (e.g. OpenAI, Google Gemini) user message can consists of multiple parts (usually to support multimodal inputs), so `UserMessage` in LC4j consists of a `List<Content> contents`.

Reply:
Sorry I get it now. fixed.